src/Simulation/DocumentationGenerationStep/ProcessDocComments.cs

```diff
@@ -0,0 +1,228 @@
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Quantum.QsCompiler;
+using Microsoft.Quantum.QsCompiler.DataTypes;
+using Microsoft.Quantum.QsCompiler.Documentation;
+using Microsoft.Quantum.QsCompiler.SyntaxTokens;
+using Microsoft.Quantum.QsCompiler.SyntaxTree;
+using Microsoft.Quantum.QsCompiler.Transformations.Core;
+using Newtonsoft.Json.Linq;
+
+#nullable enable
+
+namespace Microsoft.Quantum.Documentation
+{
+    public class ProcessDocComments
+    : SyntaxTreeTransformation<ProcessDocComments.TransformationState>
+    {
+        public class TransformationState
+        { }
+
+        private string? OutputPath;
+
+        public ProcessDocComments(
+            string? outputPath = null
+        )
+        : base(new TransformationState())
+        {
+            OutputPath = outputPath;
+            // If the output path is not null, make sure the directory exists.
+            if (outputPath != null && !Directory.Exists(outputPath))
+            {
+                Directory.CreateDirectory(outputPath);
+            }
+
+            // We provide our own custom namespace transformation, and expression kind transformation.
+            this.Namespaces = new ProcessDocComments.NamespaceTransformation(this, outputPath);
+        }
+
+        public QsCompilation OnCompilation(QsCompilation compilation) =>
+            new QsCompilation(
+                compilation.Namespaces
+                    .Select(this.Namespaces.OnNamespace)
+                    .ToImmutableArray(),
+                compilation.EntryPoints
+            );
+
+        private class NamespaceTransformation
+        : NamespaceTransformation<TransformationState>
+        {
+            private string? outputPath;
+            internal NamespaceTransformation(ProcessDocComments parent, string? outputPath)
+            : base(parent)
+            { this.outputPath = outputPath; }
+            
+            private static QsCallable AddSummaryAttribute(QsCallable callable, string summary) =>
+                callable.AddAttribute(
+                    new QsDeclarationAttribute(
+                        QsNullable<UserDefinedType>.NewValue(
+                            new UserDefinedType(
+                                NonNullable<string>.New("Microsoft.Quantum.Documentation"),
+                                NonNullable<string>.New("Summary"),
+                                QsNullable<Tuple<QsPositionInfo, QsPositionInfo>>.Null
+                            )
+                        ),
+                        SyntaxGenerator.StringLiteral(NonNullable<string>.New(summary), ImmutableArray<TypedExpression>.Empty),
+                        callable.Location.Item.Offset,
+                        QsComments.Empty
+                    )
+                );
+
+            private async Task MaybeWriteOutput(QsCustomType type, DocComment docComment)
```

Comment:
I would put these writers etc into a separate file.

Reply:
Sounds good, I agree; this was convenient when quickly prototyping, but is not the most maintainable.